drv/lpc55-sprot-server/src/main.rs

```diff
@@ -175,280 +114,618 @@ fn main() -> ! {
+                            == State::Read(ReadState::WaitingForRequest)
+                        {
+                            // Wait for interrupt
+                            break;
+                        }
+                    }
+                    ReadState::InFrame => {
+                        self.read_until_csn_deasserted(&mut rx_msg)
+                    }
+                    ReadState::FrameRead => {
+                        self.mark_busy();
+                        return handler.handle(rx_msg, tx_msg, &mut self.stats);
+                    }
+                    ReadState::Flush | ReadState::UnexpectedCsnAssert => {
+                        return None;
+                    }
+                    ReadState::Overrun => {
+                        self.mark_busy();
+                        return Some(handler.flow_error(tx_msg));
+                    }
+                    ReadState::SpProtocolCapabilitiesMismatch => {
+                        self.mark_busy();
+                        return Some(handler.protocol_error(tx_msg));
+                    }
                 }
+            }
+        }
+    }
+
+    // We are waiting for CSn to be asserted so we can transmit our response.
+    fn write_wait_for_csn_asserted(&mut self) {
+        // Get frame start/end interrupt from intstat (SSA/SSD).
+        let intstat = self.spi.intstat();
+        ringbuf_entry!(Trace::Intstat(intstat.bits()));
+
+        // CSn asserted by the SP.
+        if intstat.ssa().bit() {
+            self.spi.ssa_clear();
+            self.state = State::Write(WriteState::Writing);
+        }
+    }
+
+    fn write_response(&mut self, tx_iter: &mut dyn Iterator<Item = u8>) {
+        let mut bytes_read: usize = 0;
+        let mut state = self.state.write();
+        let mut overrun_seen = false;
+
+        // Set to true when `tx_iter` has been exhausted. Note that we may
+        // continue sending `0` after this if we are still clocking in bytes
+        // from the SP.
+        let mut response_sent = false;
 
-                // Track Spi Select Asserted and Deasserted to determine if
-                // we are in frame and update Rx/Tx state as needed.
-                if intstat.ssa().bit() {
-                    self.spi.ssa_clear();
-                    inframe = true;
+        loop {
+            let fifostat = self.spi.fifostat();
+
+            // We check for CSn deasserted at the beginning of the loop because
+            // we always want an opportunity drain our fifos after we see the bit set.
+            let csn_deasserted = self.spi.ssd();
+
+            if fifostat.txerr().bit() {
+                // We missed our window to send :(
+                self.spi.txerr_clear();
+                self.stats.tx_underrun = self.stats.tx_underrun.wrapping_add(1);
+                // If we've already written our full response, underrun doesn't matter.
+                // This can happen if we clock out BUF_SIZE of data, but are waiting
```

Comment:
This shouldnt' actually be possible, unless the SP is confused.

Reply:
The SP cannot be confused. It is the boss in this relationship.

Reply:
I guess you can argue with the wording, but fundamentally there are limits to the protocol. The SP should not be sending more than `BUF_SIZE` bytes of data in a SPI transaction.